### PR TITLE
fix: ignore __type when deserializing union for AWS Json protocols

### PR DESCRIPTION
**Pull request by 0marperez**
## Issue \#

closes [aws-sdk-kotlin#1044](https://github.com/awslabs/aws-sdk-kotlin/issues/1044)

## Description of changes

-Added `IgnoreKey` trait and adapted `JsonDeserializer` to it 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SdkFieldDescriptor.kt [hunk 1]
```diff
@@ -6,23 +6,6 @@ package aws.smithy.kotlin.runtime.serde
...
 
-/**
- * This tag interface provides a mechanism to attach type-specific metadata to any field.
- * See [aws.smithy.kotlin.runtime.serde.xml.XmlList] for an example implementation.
- *
- * For example, to specify that a list should be serialized in XML such that values are wrapped
- * in a tag called "boo", pass an instance of XmlList to the FieldDescriptor of `XmlList(elementName="boo")`.
- */
-@InternalApi
-public interface FieldTrait
```
Comment (aajtodd) [inline runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SdkFieldDescriptor.kt:17]: Why are these moved?

File: runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Indicates to deserializers to ignore field/key
+ *
+ * @param key The key to ignore in the payload
+ * @param regardlessOfInModel If true will ignore key even though the model indicates key should be there
+ *
+ */
+@InternalApi
+public data class IgnoreKey(public val key: String, public val regardlessOfInModel: Boolean = true) : FieldTrait
```
Comment (aajtodd) [inline runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: This should be in `serde-json` as it only applies there right? (see `JsonFieldTraits.kt`)

Comment (aajtodd) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: **Question**: Why do we need `regardlessOfInModel`, I would assume if we set `IgnoreKey` as a trait on the object descriptor it's going to be ignored regardless of whether it's modeled. This seems unnecessary to me but I'd like to understand what the motivation was.

File: runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SdkFieldDescriptor.kt (hunk 1 above)
Comment (0marperez) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SdkFieldDescriptor.kt:17]: Since the file was named `SdkFieldDescriptor` I thought it would make more sense to have field traits in a different place, I can change it to how it was though

File: runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt (hunk 2 above)
Comment (0marperez) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: I was thinking of a case where we have a union with a member called `__type`. If that happens we would not deserialize that member, because we're ignoring that key regardless of what the model says. The default is true anyway so it behaves as expected unless set explicitly to false.

Comment (0marperez) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: I didn't put it in a JSON specific file because I think this might be useful in the future for other deserializers like the XML one. I don't know if we have a use case already but that was my reasoning behind it.

Comment (aajtodd) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: > I was thinking of a case where we have a union with a member called __type. If that happens we would not deserialize that member, because we're ignoring that key regardless of what the model says

Shouldn't this just be a choice by codegen though? If we want it deserialized because it's in the model don't add the `IgnoreKey` trait.

Comment (0marperez) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: That makes sense but the issue this is trying to solve: https://github.com/smithy-lang/smithy/pull/1945
is that sometimes a unnecessary `__type` is serialized. What you're suggesting is to always ignore `__type`. But there could be a cases where `__type` could be in the model and we would be ignoring it because sometimes there's an extra one there unnecessarily

Comment (aajtodd) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: > What you're suggesting is to always ignore __type. But there could be a cases where __type could be in the model and we would be ignoring it because sometimes there's an extra one there unnecessarily

I don't follow, if `__type` is in the model and we _want_ to deserialize it then just don't emit the `IgnoreKey` trait during codegen. What am I missing here?

Comment (0marperez) [reply runtime/serde/common/src/aws/smithy/kotlin/runtime/serde/SerdeTraits.kt:28]: I guess that logic could live in codegen, now it's in the deserializer

File: runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+import aws.smithy.kotlin.runtime.serde.SdkFieldDescriptor
+import aws.smithy.kotlin.runtime.serde.SdkObjectDescriptor
+import aws.smithy.kotlin.runtime.serde.SerialKind
+import aws.smithy.kotlin.runtime.serde.deserializeStruct
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class JsonDeserializerIgnoresKeysTest {
+    class IgnoresKeysTest {
```
Comment (lauzadis) [inline runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt:16]: this double-nested class seems unnecessary, I think you can just declare these at the top-level without using a `companion object` for simplicity

File: runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        assertEquals(2, y)
+        assertEquals(null, z)
+    }
+
+    class IgnoresMultipleKeysTest {
+        companion object {
+            val W_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("w"))
+            val X_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("x"))
+            val Y_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("y"))
+            val Z_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("z"))
```
Comment (lauzadis) [inline runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt:127]: duplicated, can re-use the others descriptors above

**Review by lauzadis — APPROVED** [review]

File: runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonDeserializer.kt [hunk 5]
```diff
@@ -198,7 +198,13 @@ private class JsonFieldIterator(
                 val token = reader.nextTokenOf<JsonToken.Name>()
                 val propertyName = token.value
                 val field = descriptor.fields.find { it.serialName == propertyName }
-                field?.index ?: Deserializer.FieldIterator.UNKNOWN_FIELD
+
+                if (descriptor.traits.contains(IgnoreKey(propertyName))) {
```
Comment (ianbotsf) [inline runtime/serde/serde-json/common/src/aws/smithy/kotlin/runtime/serde/json/JsonDeserializer.kt:202]: **Style**: Consider `if (IgnoreKey(propertyName) in descriptor.traits)`

File: runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+import aws.smithy.kotlin.runtime.serde.deserializeStruct
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class JsonDeserializerIgnoresKeysTest {
+    private val X_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("x"))
+    private val Y_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("y"))
+    private val Z_DESCRIPTOR = SdkFieldDescriptor(SerialKind.Integer, JsonSerialName("z"))
+    private val OBJ_DESCRIPTOR = SdkObjectDescriptor.build {
+        trait(IgnoreKey("z")) // <----
```
Comment (ianbotsf) [inline runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt:20]: **Question**: What does the comment `// <----` mean? Same question on L120-121.

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/serde/JsonParserGenerator.kt [hunk 7]
```diff
@@ -24,6 +24,13 @@ open class JsonParserGenerator(
 
     open val defaultTimestampFormat: TimestampFormatTrait.Format = TimestampFormatTrait.Format.EPOCH_SECONDS
 
+    open fun descriptorGenerator(
+        ctx: ProtocolGenerator.GenerationContext,
+        shape: Shape,
+        members: List<MemberShape>,
+        writer: KotlinWriter,
+    ): JsonSerdeDescriptorGenerator = JsonSerdeDescriptorGenerator(ctx.toRenderingContext(protocolGenerator, shape, writer), members, supportsJsonNameTrait)
```
Comment (ianbotsf) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/serde/JsonParserGenerator.kt:32]: **Style**: Line length > 120 chars

**Review by ianbotsf — APPROVED** [review]
Looks good overall!

File: runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+package aws.smithy.kotlin.runtime.serde.json
+
+import aws.smithy.kotlin.runtime.serde.SdkFieldDescriptor
+import aws.smithy.kotlin.runtime.serde.SdkObjectDescriptor
+import aws.smithy.kotlin.runtime.serde.SerialKind
+import aws.smithy.kotlin.runtime.serde.deserializeStruct
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class JsonDeserializerIgnoresKeysTest {
```
Comment (aajtodd) [inline runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerIgnoresKeysTest.kt:15]: **Suggestion**: Every test has explicitly modeled field descriptors. You're missing coverage for unmodeled fields that we want to skip rather than enumerate as unknown.

**Review by aajtodd — APPROVED** [review]
Looks good, one suggestion but otherwise approved.
